storage/libradb/src/lib.rs

```diff
@@ -620,16 +620,29 @@ impl DbReader for LibraDB {
         limit: u64,
     ) -> Result<Vec<(u64, ContractEvent)>> {
         gauged_api("get_events", || {
+            let events_with_proofs = self.get_events_with_proofs(event_key, start, order, limit)?;
+            let events = events_with_proofs
+                .into_iter()
+                .map(|e| (e.transaction_version, e.event))
+                .collect();
+            Ok(events)
+        })
+    }
+
+    fn get_events_with_proofs(
+        &self,
+        event_key: &EventKey,
+        start: u64,
+        order: Order,
+        limit: u64,
+    ) -> Result<Vec<EventWithProof>> {
+        gauged_api("get_events_with_proofs", || {
             let version = self
```

Comment:
I think the version against which the proof is provided should be passed in -- so that proofs within the same client request are against a same latest version that the client knows about.

Reply:
This is a good UX question which I've been trying to figure out.
Currently the client is expected to batch its call with a get_state_proof call like it's done here in the integration test: it allows the client to upgrade its state to the latest state (and thus the latest set of validators in case of an epoch change) and to verify the proofs.

You can see in the integration test how I've done it. (Although it's a bit hack-ish since we don't have a state with the validator set in the integration tests...)

Notice this is the same as I've done for the transaction_with_proofs UX, but this is maybe terrible UX and I should instead have the client specify the version for which it wants to get proofs, indeed.

If we agree that it's better to have the client specify the version for which it wants the proofs, then I should change the recent `get_transactions_with_proofs` API too:
https://github.com/libra/libra/blob/8332031c482cdbc678b16310971e8572a2cfb995/json-rpc/src/methods.rs#L397-L415 

What do you think @xli  ? On my side I think this would be cleaner if the client could specify its desired version, indeed since then it allows to have non-batched calls with the API :)

Reply:
There are two matters here: 
1. Client provided version or server latest version. Allowing the client to optionally pass in a version, defaulting to the server latest version is fine.
2. all batched queries in the same request should return a consistent view. i.e. add `ledger_version` param to the DB method you add. Check out other DB methods use it, for example `get_transactions()` -- it should not return things beyond that version, and the proofs should be relative to that version.  (I think in your case the underlying `get_events_by_event_key()` has already dealt with that.

Reply:
Done, we now support both batched and non-batched queries thanks to a `known_version` optional parameter